python/tank/util/shotgun.py

```diff
@@ -1141,9 +1147,79 @@ def _create_published_file(tk, context, path, name, version_number, task, commen
     """
     Creates a publish entity in shotgun given some standard fields.
+
+    :param tk: :class:`~sgtk.Sgtk` instance
+    :param context: A :class:`~sgtk.Context` to associate with the publish. This will
+                    populate the task and entity link in Shotgun.
+    :param path: The path to the file or sequence we want to publish. If the
+                 path is a sequence path it will be abstracted so that
+                 any sequence keys are replaced with their default values.
+    :param name: A name, without version number, which helps distinguish
+               this publish from other publishes. This is typically
+               used for grouping inside of Shotgun so that all the
+               versions of the same "file" can be grouped into a cluster.
+               For example, for a maya publish, where we track only
+               the scene name, the name would simply be that: the scene
+               name. For something like a render, it could be the scene
+               name, the name of the AOV and the name of the render layer.
+    :param version_number: The version number of the item we are publishing.
+    :param task: Shotgun Task dictionary to assciate with publish or none
+    :param comment: Comments string to associate with publish
+    :param published_file_type: Shotgun publish type dictionary to
+                associate with publish
+    :param created_by_user: User entity to associate with publish or None
+                if current user should be used.
+    :param created_at: timestamp to associate with publish or None for default.
+    :param version_entity: Version dictionary to associate with publish or None.
+    :param sg_fields: dictionary of additional data to add to publish.
+
+    :returns: the result of the shotgun API create method.
     """
+
+    data = {
+        "description": comment,
+        "name": name,
+        "project": context.project,
+        "task": task,
+        "version_number": version_number,
+        }
+
+    # we set the optional additional fields first so we don't allow overwriting the standard parameters
+    if sg_fields is None:
+        sg_fields = {}
+    data.update(sg_fields)
+
+    if created_by_user:
+        data["created_by"] = created_by_user
+    else:
+        # use current user
+        sg_user = login.get_current_user(tk)
+        if sg_user:
+            data["created_by"] = sg_user
+
+    if created_at:
+        data["created_at"] = created_at
+
     published_file_entity_type = get_published_file_entity_type(tk)
 
+    if published_file_type:
+        if published_file_entity_type == "PublishedFile":
+            data["published_file_type"] = published_file_type
+        else:
+            # using legacy type TankPublishedFile
+            data["tank_type"] = published_file_type
+
+    if version_entity:
+        data["version"] = version_entity
+
+    # if the context does not have an entity, link it up to the project
```

Comment:
Could you elaborate in the comments why this is an acceptable situation and why the fallback is appropriate?